### PR TITLE
Added more robust `_GIT_URI_REGEX`

### LINKED ISSUE #2824: [BUG] Git url regex matches any windows path in projects.run()
Thank you for submitting an issue. Please refer to our [issue policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md) for additional information about bug reports. For help with debugging your code, please refer to [Stack Overflow](https://stackoverflow.com/questions/tagged/mlflow).

**Please fill in this bug report template to ensure a timely and thorough response.**

### Willingness to contribute
The MLflow Community encourages bug fix contributions. Would you or another member of your organization be willing to contribute a fix for this bug to the MLflow code base?

- [x] Yes. I can contribute a fix for this bug independently.
- [ ] Yes. I would be willing to contribute a fix for this bug with guidance from the MLflow community.
- [ ] No. I cannot contribute a bug fix at this time.

### System information
- **Have I written custom code (as opposed to using a stock example script provided in MLflow)**: No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Windows 10
- **MLflow installed from (source or binary)**: binary
- **MLflow version (run ``mlflow --version``)**: 1.7.2
- **Python version**: Python 3.7.3
- **npm version, if running the dev UI**: N/A
- **Exact command to reproduce**: python -c "from mlflow import projects;projects.run('C:\asd')"

### Describe the problem
Describe the problem clearly here. Include descriptions of the expected behavior and the actual behavior.

Git uri regex matches absolute Windows path starting with drive (e.g. C:\..). As a result running projects.run('C:\...') results in the error:
```
git.exc.GitCommandError: Cmd('git') failed due to: exit code(128)
  cmdline: git fetch -v origin
  stderr: 'fatal: 'C:/asd' does not appear to be a git repository
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.'
```

The git uri regex seemse to be too broad, matching any path containing `:`
```
_GIT_URI_REGEX = re.compile(r"^[^/]*:")
```

### Code to reproduce issue
Provide a reproducible test case that is the bare minimum necessary to generate the problem.

```
from mlflow import projects
projects.run('C:\\asd')
```

### Other info / logs
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

```
>>> from mlflow import projects
>>> projects.run('C:\\asd')
2020/05/14 09:45:26 INFO mlflow.projects: === Fetching project from C:\asd into C:\Users\USER_NAME\AppData\Local\Temp\tmpciv1cczr ===
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "D:\miniconda3\lib\site-packages\mlflow\projects\__init__.py", line 288, in run
    use_conda=use_conda, storage_dir=storage_dir, synchronous=synchronous, run_id=run_id)
  File "D:\miniconda3\lib\site-packages\mlflow\projects\__init__.py", line 104, in _run
    work_dir = _fetch_project(uri=uri, force_tempdir=False, version=version)
  File "D:\miniconda3\lib\site-packages\mlflow\projects\__init__.py", line 342, in _fetch_project
    _fetch_git_repo(parsed_uri, version, dst_dir)
  File "D:\miniconda3\lib\site-packages\mlflow\projects\__init__.py", line 382, in _fetch_git_repo
    origin.fetch()
  File "D:\miniconda3\lib\site-packages\git\remote.py", line 790, in fetch
    res = self._get_fetch_info_from_stderr(proc, progress)
  File "D:\miniconda3\lib\site-packages\git\remote.py", line 674, in _get_fetch_info_from_stderr
    proc.wait(stderr=stderr_text)
  File "D:\miniconda3\lib\site-packages\git\cmd.py", line 412, in wait
    raise GitCommandError(self.args, status, errstr)
git.exc.GitCommandError: Cmd('git') failed due to: exit code(128)
  cmdline: git fetch -v origin
  stderr: 'fatal: 'C:/asd' does not appear to be a git repository
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.'
```

### What component(s), interfaces, languages, and integrations does this bug affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for
Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [x] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [x] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations

### LINKED ISSUE #5998: [BUG] mlflow.projects.run can't recognize local URI
### Willingness to contribute

Yes. I can contribute a fix for this bug independently.

### MLflow version

1.26.1

### System information

- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Windows 10
- **Python version**: 3.10.4
- **yarn version, if running the dev UI**: N/A

### Describe the problem

I'm recreating the hyperparameter optimization project from https://github.com/mlflow/mlflow/tree/master/examples/hyperparam , i.e. I have coded all the files from scratch on my local machine, and I am running into the following problem while invoking the `search_random.py` script using the command `mlflow -e random --experiment-id 2 . --env-manager local` from the root directory of the project.

The error:
```python-traceback
File "D:\Softwares\Anaconda\envs\mlflow\lib\site-packages\git\cmd.py", line 502, in wait
    raise GitCommandError(remove_password_if_present(self.args), status, errstr)
git.exc.GitCommandError: Cmd('git') failed due to: exit code(128)
  cmdline: git fetch -v --depth=1 origin
  stderr: 'fatal: 'D:/MyStuff/Learn/MachineLearning/MLFlow/LearnMLFlow/tut2' does not appear to be a git repository
fatal: Could not read from remote repository.'
```

Through some debugging, I found that the exception is thrown at the point in the code where run an instance of the `train.py` script on the new thread:

```python
with mlflow.start_run(nested=True) as child_run:
    p = mlflow.projects.run(
        run_id=child_run.info.run_id,
        uri=".",
        entry_point="train",
        parameters={
            "training_data": training_data,
            "epochs": str(nepochs),
            "learning_rate": str(lr),
            "momentum": str(momentum),
            "seed": str(seed),
        },
        experiment_id=experiment_id,
        synchronous=False,
        env_manager='local'
    )
    succeeded = p.wait()
```

Specifically, the exception is thrown on the `p.wait()` call. I can't seem to understand why MLFlow thinks the project is at a remote repository even though I've specified the `uri` for the project as the current directory in the constructor. I've also tried replacing the `.` with the absolute path of the project directory, but the error persists.

### Tracking information

_No response_

### Code to reproduce issue

```python
with mlflow.start_run(nested=True) as child_run:
    p = mlflow.projects.run(
        run_id=child_run.info.run_id,
        uri=".",
        entry_point="train",
        parameters={
            "training_data": training_data,
            "epochs": str(nepochs),
            "learning_rate": str(lr),
            "momentum": str(momentum),
            "seed": str(seed),
        },
        experiment_id=experiment_id,
        synchronous=False,
        env_manager='local'
    )
    succeeded = p.wait()
```

### Other info / logs

```python-traceback
2022/05/31 03:07:45 INFO mlflow.projects.utils: === Created directory C:\Users\dhruv\AppData\Local\Temp\tmphzol9g4d for downloading remote URIs passed to arguments of type 'path' ===
2022/05/31 03:07:45 INFO mlflow.projects.backend.local: === Running command 'python search_random.py  http://archive.ics.uci.edu/ml/machine-learning-databases/wine-quality/winequality-white.csv --max-runs 8 --max-p 2 --epochs 32 --metric rmse --seed 97531' in run with ID '6e2312ed5b914f32b933f160035dc3cf' ===
2022/05/31 03:07:48 INFO mlflow.projects.utils: === Fetching project from D:\MyStuff\Learn\MachineLearning\MLFlow\LearnMLFlow\tut2 into C:\Users\dhruv\AppData\Local\Temp\tmpg0sm6lo5 ===
Traceback (most recent call last):
  File "D:\MyStuff\Learn\MachineLearning\MLFlow\LearnMLFlow\tut2\search_random.py", line 123, in <module>
    run()
  File "D:\Softwares\Anaconda\envs\mlflow\lib\site-packages\click\core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "D:\Softwares\Anaconda\envs\mlflow\lib\site-packages\click\core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "D:\Softwares\Anaconda\envs\mlflow\lib\site-packages\click\core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "D:\Softwares\Anaconda\envs\mlflow\lib\site-packages\click\core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "D:\MyStuff\Learn\MachineLearning\MLFlow\LearnMLFlow\tut2\search_random.py", line 90, in run
    _, null_train_loss, null_val_loss, null_test_loss = new_eval(0, experiment_id)((0, 0))
  File "D:\MyStuff\Learn\MachineLearning\MLFlow\LearnMLFlow\tut2\search_random.py", line 47, in eval
    p = mlflow.projects.run(
  File "D:\Softwares\Anaconda\envs\mlflow\lib\site-packages\mlflow\projects\__init__.py", line 331, in run
    submitted_run_obj = _run(
  File "D:\Softwares\Anaconda\envs\mlflow\lib\site-packages\mlflow\projects\__init__.py", line 95, in _run
    submitted_run = backend.run(
  File "D:\Softwares\Anaconda\envs\mlflow\lib\site-packages\mlflow\projects\backend\local.py", line 64, in run
    work_dir = fetch_and_validate_project(project_uri, version, entry_point, params)
  File "D:\Softwares\Anaconda\envs\mlflow\lib\site-packages\mlflow\projects\utils.py", line 130, in fetch_and_validate_project
    work_dir = _fetch_project(uri=uri, version=version)
  File "D:\Softwares\Anaconda\envs\mlflow\lib\site-packages\mlflow\projects\utils.py", line 164, in _fetch_project
    _fetch_git_repo(parsed_uri, version, dst_dir)
  File "D:\Softwares\Anaconda\envs\mlflow\lib\site-packages\mlflow\projects\utils.py", line 202, in _fetch_git_repo
    origin.fetch(depth=GIT_FETCH_DEPTH)
  File "D:\Softwares\Anaconda\envs\mlflow\lib\site-packages\git\remote.py", line 886, in fetch
    res = self._get_fetch_info_from_stderr(proc, progress,
  File "D:\Softwares\Anaconda\envs\mlflow\lib\site-packages\git\remote.py", line 750, in _get_fetch_info_from_stderr
    proc.wait(stderr=stderr_text)
  File "D:\Softwares\Anaconda\envs\mlflow\lib\site-packages\git\cmd.py", line 502, in wait
    raise GitCommandError(remove_password_if_present(self.args), status, errstr)
git.exc.GitCommandError: Cmd('git') failed due to: exit code(128)
  cmdline: git fetch -v --depth=1 origin
  stderr: 'fatal: 'D:/MyStuff/Learn/MachineLearning/MLFlow/LearnMLFlow/tut2' does not appear to be a git repository
fatal: Could not read from remote repository.'
2022/05/31 03:07:49 ERROR mlflow.cli: === Run (ID '6e2312ed5b914f32b933f160035dc3cf') failed ===
```

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [X] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [X] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [X] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### PR DESCRIPTION
**Pull request by dkapur17**
## Related Issues/PRs

Fix #5998
Close #2824
Close #2825
Close #3274

## What changes are proposed in this pull request?

Changed the regular expression used to match Git repositories in `mlflow/projects/utils.py` to a more robust one.

## How is this patch tested?

Manually tested by using the functionality in a script.

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
2. Click `Details` on the right to open the job page of CircleCI.
3. Click the `Artifacts` tab.
4. Click `docs/build/html/index.html`.
5. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [x] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [x] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [x] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

Signed-off-by: Dhruv Kapur <dhruvkapur4@gmail.com>

### CODE REVIEW

**Comment by harupy** [comment]
@dkapur17 Thanks for the PR, can you revert unrelated formatting changes?

**Comment by dkapur17** [comment]
> @dkapur17 Thanks for the PR, can you revert unrelated formatting changes?

Will do.

File: mlflow/projects/utils.py [hunk 1]
```diff
@@ -353,4 +352,4 @@ def get_databricks_env_vars(tracking_uri):
         env_vars["DATABRICKS_TOKEN"] = config.token
     if config.ignore_tls_verification:
         env_vars["DATABRICKS_INSECURE"] = str(config.ignore_tls_verification)
-    return env_vars
+    return env_vars
```
Comment (harupy) [inline mlflow/projects/utils.py:355]: Can we revert this change?

File: mlflow/projects/utils.py [hunk 2]
```diff
@@ -29,8 +29,7 @@
 )
 from mlflow.utils.rest_utils import augmented_raise_for_status
 
-# TODO: this should be restricted to just Git repos and not S3 and stuff like that
-_GIT_URI_REGEX = re.compile(r"^[^/]*:")
+_GIT_URI_REGEX = re.compile(r"((git|ssh|http(s)?)|(git@[\w\.]+))(:(//)?)([\w\.@\:/\-~]+)(\.git)(/)?")
```
Comment (harupy) [inline mlflow/projects/utils.py:32]: Can we add a test (e.g. `test_git_uri_regex`) in `tests/projects/test_utils.py` to ensure this doesn't match a windows path?

**Comment by dkapur17** [comment]
autoformat

**Comment by dkapur17** [comment]
`autoformat`

**Comment by harupy** [comment]
Some tests failed:

https://github.com/mlflow/mlflow/runs/6668747346?check_suite_focus=true#step:9:1148

```
E           mlflow.exceptions.ExecutionException: Could not find subdirectory  of file:///tmp/pytest-of-runner/pytest-0/test__fetch_project0/git_repo
```

The new Git URI pattern doesn't match `file:///tmp/pytest-of-runner/pytest-0/test__fetch_project0/git_repo`.

**Comment by dkapur17** [comment]
@harupy yeah, I believe this might work, though I've never seen a git repo URI like `file:///tmp/pytest-of-runner/pytest-0/test__fetch_project0/git_repo`. This could simply be a file that is named `git_repo`, or am I missing something here?

**Comment by harupy** [comment]
@dkapur17 

> This could simply be a file that is named git_repo

Yes, but this could be a directory named `git_repo` that's a git repository.

**Comment by dkapur17** [comment]
Ah, I see. Then yes, your proposed solution should work correctly.

**Comment by harupy** [comment]
I think the failed tests should pass with this change.

File: mlflow/projects/utils.py [hunk 3]
```diff
@@ -97,10 +99,32 @@ def _is_file_uri(uri):
...
+
+
 def _is_local_uri(uri):
-    """Returns True if the passed-in URI should be interpreted as a path on the local filesystem."""
-    return not _GIT_URI_REGEX.match(uri)
+    """Returns True if passed-in URI should be interpreted as a path on the local filesystem."""
+    if _GIT_URI_REGEX.match(uri):
+        return False
 
+    pased_uri = urllib.parse.urlparse(uri).scheme
```
[the remark this answers is not in the document]
Comment (harupy) [reply mlflow/projects/utils.py:118]: @dkapur17 Can you fix this?

File: mlflow/projects/utils.py [hunk 4]
```diff
@@ -97,10 +99,32 @@ def _is_file_uri(uri):
...
+
+
 def _is_local_uri(uri):
-    """Returns True if the passed-in URI should be interpreted as a path on the local filesystem."""
-    return not _GIT_URI_REGEX.match(uri)
+    """Returns True if passed-in URI should be interpreted as a path on the local filesystem."""
+    if _GIT_URI_REGEX.match(uri):
+        return False
 
+    parsed_uri = urllib.parse.urlparse(uri).scheme
```
Comment (harupy) [inline mlflow/projects/utils.py:118]:
```suggestion
    parsed_uri = urllib.parse.urlparse(uri)
```

Sorry, my code suggestion was wrong. We need to remove `.scheme`.

Comment (harupy) [reply mlflow/projects/utils.py]: @dkapur17 I pushed https://github.com/mlflow/mlflow/pull/6000/commits/e7aab7fba4eebbf3c77b06027c8fd70f6408364e to address this comment :)

Comment (dkapur17) [reply mlflow/projects/utils.py]: Awesome!

**Review by harupy — APPROVED** [review]
LGTM :)

**Comment by jaebjgh** [comment]
I have a problem, which might be related to this. 
When starting an MLproject with "mlflow run --experiment-name "machine_speed_forecasting" . " in mlflow at version 1.26, it runs my code inside the Docker Image, which i specified inside the MLproject file. The MLproject file is located directly where the line is executed, hence the "." indicating to look in the present directory. 
The first line states the following: 

> INFO mlflow.projects.docker: === Building docker image forecasting_pipeline:77fab66 ===

This behaviour is perfectly fine and working just as I thought it would.

**But** when running the exact same line after upgrading MLflow to 1.27, it doesn't start the local MLproject file, but instead tries to find a Git Repository and throws an error. 
The first line here is as follows:

> INFO mlflow.projects.utils: === Fetching project from . into /tmp/tmpap9m1q3c ===

> cmdline: git fetch -v --depth=1 origin (unknown)
>   stderr: 'fatal: couldn't find remote ref (unknown)'

So maybe there is something wrong with the url-parsing or it doesn't check the case where someone just uses the "." to indicate the directory?

**Comment by pikulmar** [comment]
> I have a problem, which might be related to this. When starting an MLproject with
> 
> > mlflow run --experiment-name "machine_speed_forecasting" .
> 
> in mlflow 1.26, it runs my code inside the Docker Image, which i specified inside the MLproject file. The MLproject file is located directly where the line is executed, hence the "." indicating to look in the present directory. The first line states the following:
> 
> > INFO mlflow.projects.docker: === Building docker image forecasting_pipeline:77fab66 ===
> 
> This behaviour is perfectly fine and working just as I thought it would.
> 
> **But** when running the exact same line after upgrading MLflow to 1.27, it doesn't start the local MLproject file, but instead tries to find a Git Repository and throws an error. The first line here is as follows:
> 
> > INFO mlflow.projects.utils: === Fetching project from . into /tmp/tmpap9m1q3c ===
> 
> > cmdline: git fetch -v --depth=1 origin (unknown)
> > stderr: 'fatal: couldn't find remote ref (unknown)'
> 
> So maybe there is something wrong with the url-parsing or it doesn't check the case where someone just uses the "." to indicate the directory?

I can confirm the issue with v1.27.0. As of 043a54d74b451edfca72ed03b749a3594ade469f, the issue appears to be resolved.

**Pull request merged**